<a href="https://colab.research.google.com/github/eya-methnani/Transfer-Learning-Notebook-COURSE/blob/main/transfer_learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:

!pip install tensorflow keras


In [ ]:
from keras.applications.vgg16 import VGG16
# load model
model = VGG16()
# summarize the model
model.summary()

Model: "vgg16"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ input_layer_3 (InputLayer)           │ (None, 224, 224, 3)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv1 (Conv2D)                │ (None, 224, 224, 64)        │           1,792 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_conv2 (Conv2D)                │ (None, 224, 224, 64)        │          36,928 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block1_pool (MaxPooling2D)           │ (None, 112, 112, 64)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv1 (Conv2D)                │ (None, 112, 112, 128)       │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_conv2 (Conv2D)                │ (None, 112, 112, 128)       │         147,584 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block2_pool (MaxPooling2D)           │ (None, 56, 56, 128)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv1 (Conv2D)                │ (None, 56, 56, 256)         │         295,168 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv2 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_conv3 (Conv2D)                │ (None, 56, 56, 256)         │         590,080 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block3_pool (MaxPooling2D)           │ (None, 28, 28, 256)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv1 (Conv2D)                │ (None, 28, 28, 512)         │       1,180,160 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv2 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_conv3 (Conv2D)                │ (None, 28, 28, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block4_pool (MaxPooling2D)           │ (None, 14, 14, 512)         │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv1 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv2 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_conv3 (Conv2D)                │ (None, 14, 14, 512)         │       2,359,808 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ block5_pool (MaxPooling2D)           │ (None, 7, 7, 512)           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 25088)               │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ fc1 (Dense)                          │ (None, 4096)                │     102,764,544 │
├──────────────────────────────────────┼─────────────────────────────┼──────────────

 Total params: 138,357,544 (527.79 MB)

 Trainable params: 138,357,544 (527.79 MB)

 Non-trainable params: 0 (0.00 B)

**VGG16 Model Summary:**

* **Input Layer:**

    input_layer_3 (InputLayer): Accepts images of size 224x224x3 (RGB format).
* **Convolutional and Pooling Layers:**

    Block 1:

    * block1_conv1 and block1_conv2: Two convolutional layers with 64 filters each, extracting features from the image.

    * block1_pool: A max pooling layer that reduces the spatial dimensions by half, from 224x224 to 112x112.

    Block 2:

    * block2_conv1 and block2_conv2: Two convolutional layers with 128 filters each, capturing more complex features.

    * block2_pool: Max pooling further reduces the dimensions to 56x56.

    Block 3:

    * block3_conv1, block3_conv2, block3_conv3: Three convolutional layers with 256 filters each.

    * block3_pool: Max pooling reduces the dimensions to 28x28.

    Block 4:

    * block4_conv1, block4_conv2, block4_conv3: Three convolutional layers with 512 filters.

    * block4_pool: Reduces the dimensions to 14x14.

    Block 5:

    * block5_conv1, block5_conv2, block5_conv3: Three convolutional layers with 512 filters.

    * block5_pool: Reduces dimensions to 7x7, preparing for the dense layers.

*  **Dense Layers:**

  Flatten Layer:

    * flatten: Reshapes the 7x7x512 output from the final pooling layer into a 1D vector of size 25,088.

  Fully Connected Layers:

    * fc1 and fc2: Dense layers with 4,096 units each, responsible for learning complex representations.

  Output Layer:

    * predictions: A dense layer with 1,000 units, using softmax activation for classification into 1,000 classes (e.g., ImageNet categories).


* **Summary**:

Total Parameters: 138,357,544 (527.79 MB).
Trainable Parameters: All 138,357,544 parameters are trainable by default.


In [ ]:
from keras.applications.inception_v3 import InceptionV3
# load model
model = InceptionV3()
# summarize the model
model.summary()

96112376/96112376 ━━━━━━━━━━━━━━━━━━━━ 1s 0us/step


Model: "inception_v3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer_4             │ (None, 299, 299, 3)    │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d (Conv2D)           │ (None, 149, 149, 32)   │            864 │ input_layer_4[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ batch_normalization       │ (None, 149, 149, 32)   │             96 │ conv2d[0][0]           │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ activation (Activation)   │ (None, 149, 149, 32)   │              0 │ batch_normalization[0… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_1 (Conv2D)         │ (None, 147, 147, 32)   │          9,216 │ activation[0][0]       │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ batch_normalization_1     │ (None, 147, 147, 32)   │             96 │ conv2d_1[0][0]         │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ activation_1 (Activation) │ (None, 147, 147, 32)   │              0 │ batch_normalization_1… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_2 (Conv2D)         │ (None, 147, 147, 64)   │         18,432 │ activation_1[0][0]     │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ batch_normalization_2     │ (None, 147, 147, 64)   │            192 │ conv2d_2[0][0]         │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ activation_2 (Activation) │ (None, 147, 147, 64)   │              0 │ batch_normalization_2… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling2d             │ (None, 73, 73, 64)     │              0 │ activation_2[0][0]     │
│ (MaxPooling2D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_3 (Conv2D)         │ (None, 73, 73, 80)     │          5,120 │ max_pooling2d[0][0]    │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ batch_normalization_3     │ (None, 73, 73, 80)     │            240 │ conv2d_3[0][0]         │
│ (BatchNormalization)      │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ activation_3 (Activation) │ (None, 73, 73, 80)     │              0 │ batch_normalization_3… │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_4 (Conv2D)         │ (None, 71, 71, 192)    │        138,240 │ activation_3[0][0]     │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ batch_normalization_4     │ (None, 71, 71, 192)    │            576 │ conv2d_4[0][0]         │
│ (BatchNormalization)      │                        │                │                        │
├──────────────────────

 Total params: 23,851,784 (90.99 MB)

 Trainable params: 23,817,352 (90.86 MB)

 Non-trainable params: 34,432 (134.50 KB)

**Explanation of Model Components:**

 * Input Layer:

    The default input shape is (299, 299, 3) for RGB images.

    This layer can be adjusted to accommodate different image sizes for transfer learning, as long as they meet minimum size requirements.

 * Inception Modules:

    InceptionV3 is built around the Inception module, where multiple filter sizes (1x1, 3x3, 5x5, etc.) operate in parallel. This enables the model to capture fine, medium, and coarse features simultaneously.

    Each Inception module outputs a series of feature maps concatenated along the depth, enhancing the network's ability to represent complex patterns in images.

* Auxiliary Classifier:

  A unique aspect of InceptionV3 is the addition of auxiliary classifiers at intermediate layers. These classifiers improve gradient flow and regularize the network during training.

* Pooling Layers:

  MaxPooling and AveragePooling layers are used to reduce the dimensionality of the feature maps, helping the model to learn more efficiently while retaining important features.

* Fully Connected Layers:

    At the end of the convolutional layers, the output is flattened and passed through fully connected (Dense) layers. The final Dense layer uses softmax activation to predict class probabilities.

* Total Parameters:

  The InceptionV3 model has over 20 million parameters (exact counts may vary with configurations). The default pre-trained model can be used for feature extraction or fine-tuned by adjusting the last layers for custom classification tasks.

In [ ]:
# example of loading the resnet50 model
from tensorflow.keras.applications.resnet50 import ResNet50
# load model
model = ResNet50()
# summarize the model
model.summary()

102981632/102967424 [==============================] - 42s 0us/step
Model: "resnet50"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_4 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 conv1_pad (ZeroPadding2D)      (None, 230, 230, 3)  0           ['input_4[0][0]']                
                                                                                                  
 conv1_conv (Conv2D)            (None, 112, 112, 64  9472        ['conv1_pad[0][0]']              
                                )                                                                 
                       

**Explanation of Model Components:**

* Input Layer:

  The default input shape for ResNet50 is (224, 224, 3) for RGB images.

* Residual Blocks:

  The model consists of four main stages, each containing several residual blocks.

  A residual block contains three convolutional layers, batch normalization, ReLU activations, and a shortcut connection (skip connection) that adds the input to the output, helping preserve gradients and stabilize training.

* Pooling Layers:

  The model begins with an initial Conv2D layer, followed by a MaxPooling2D layer that reduces spatial dimensions.

  GlobalAveragePooling2D is used after the final residual block, reducing the output of the convolutional layers to a 1D vector, which feeds into the Dense layer.

* Dense Layer:

  The final layer is a Dense layer with 1,000 units (for ImageNet classification) and softmax activation for predicting class probabilities.

* Total Parameters:

  ResNet50 has 25,636,712 parameters in total, with 25,583,592 trainable parameters and 53,120 non-trainable parameters (such as those related to batch normalization).



***Comme moyen de classification***

In [20]:

# load an image from file
#image = load_img('C:/Users/esprit/Desktop/cours DM/dog.jpg', target_size=(224, 224))
from keras.preprocessing.image import load_img
from keras.preprocessing.image import img_to_array
from keras.applications.vgg16 import preprocess_input
from keras.applications.vgg16 import decode_predictions
from keras.applications.vgg16 import VGG16
# load an image from file
#image = load_img('D:/Users/esprit/Desktop/cours DM/dog.jpg', target_size=(224, 224))

import requests
from PIL import Image
from io import BytesIO

url='https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQAnI33utT_AfatFvwqr9LLt599wJoIriqyIw&s'
response = requests.get(url)
image = Image.open(BytesIO(response.content))

# Resize the image if needed
image = image.resize((224, 224))
image.show()



# convert the image pixels to a numpy array
image = img_to_array(image)
# reshape data for the model
image = image.reshape((1, image.shape[0], image.shape[1], image.shape[2]))
# prepare the image for the VGG model
image = preprocess_input(image)
# load the model
model = VGG16()
# predict the probability across all output classes
yhat = model.predict(image)
# convert the probabilities to class labels
label = decode_predictions(yhat)
# retrieve the most likely result, e.g. highest probability
label = label[0][0]
# print the classification
print('%s (%.2f%%)' % (label[1], label[2]*100))

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 678ms/step
Appenzeller (62.48%)


* Load VGG16 Model and Predict:

  Load the pre-trained VGG16 model and predict the image's class.
  
  The model identified the object as an "Appenzeller" with a 62.48% confidence.

In [16]:
len(yhat[0])

1000

In [17]:
import numpy as np

# Decode all predictions
decoded_labels = decode_predictions(yhat, top=1000)[0]  # Decodes all 1000 classes

# Print top 10 predictions
for i in range(10):
    print(f"{decoded_labels[i][1]}: {decoded_labels[i][2]*100:.2f}%")


Appenzeller: 62.48%
EntleBucher: 17.44%
Greater_Swiss_Mountain_dog: 16.60%
Bernese_mountain_dog: 1.73%
Rottweiler: 0.90%
Border_collie: 0.16%
kelpie: 0.11%
Tibetan_mastiff: 0.04%
Doberman: 0.04%
beagle: 0.04%


***Comme feauture extracture***

In [22]:
from keras.preprocessing.image import load_img
from keras.preprocessing.image import img_to_array
from keras.applications.vgg16 import preprocess_input
from keras.applications.vgg16 import decode_predictions
from keras.applications.vgg16 import VGG16
from keras.models import Model
from pickle import dump
# load an image from file

#image = load_img('C:/Users/esprit/Desktop/cours DM/dog.jpg', target_size=(224, 224))

url='https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQAnI33utT_AfatFvwqr9LLt599wJoIriqyIw&s'
response = requests.get(url)
image = Image.open(BytesIO(response.content))

# Resize the image if needed
image = image.resize((224, 224))
image.show()

# convert the image pixels to a numpy array
image = img_to_array(image)
# reshape data for the model
image = image.reshape((1, image.shape[0], image.shape[1], image.shape[2]))
# prepare the image for the VGG model
image = preprocess_input(image)
# load model
model = VGG16()
# remove the output layer
model = Model(inputs=model.inputs, outputs=model.layers[-2].output)
# get extracted features
features = model.predict(image)
print(features.shape)
# save to file
dump(features, open('dog.pkl', 'wb'))

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 1s/step
(1, 4096)


This code performs feature extraction from an image using the VGG16 model. By removing the output layer, it captures high-level features from the second-to-last layer

To enable feature extraction, the model is modified to remove the final output layer (classification layer), outputting the activations from the second-to-last layer instead

The modified model is used to predict features from the input image.

This yields a feature vector of shape (1, 4096), which represents the high-level features extracted by the VGG16 model.

The feature vector is saved to a file named dog.pkl using pickle.dump, allowing it to be reused without reprocessing the image


In [ ]:
model.summary()

Model: "functional_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_5 (InputLayer)         [(None, 224, 224, 3)]     0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, 224, 224, 64)      1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, 224, 224, 64)      36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, 112, 112, 64)      0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, 112, 112, 128)     73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, 112, 112, 128)     147584    
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, 56, 56, 128)      

In [ ]:
# load model without classifier layers
from keras.applications.vgg16 import VGG16
from keras.models import Model
from keras.layers import Dense
from keras.layers import Flatten
model = VGG16(include_top=False, input_shape=(300, 300, 3))
flat1 = Flatten()(model.layers[-1].output)
class1 = Dense(1024, activation='relu')(flat1)
output = Dense(10, activation='softmax')(class1)
# define new model
model = Model(inputs=model.inputs, outputs=output)

In [ ]:
model.summary()

Model: "functional_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
input_6 (InputLayer)         [(None, 300, 300, 3)]     0         
_________________________________________________________________
block1_conv1 (Conv2D)        (None, 300, 300, 64)      1792      
_________________________________________________________________
block1_conv2 (Conv2D)        (None, 300, 300, 64)      36928     
_________________________________________________________________
block1_pool (MaxPooling2D)   (None, 150, 150, 64)      0         
_________________________________________________________________
block2_conv1 (Conv2D)        (None, 150, 150, 128)     73856     
_________________________________________________________________
block2_conv2 (Conv2D)        (None, 150, 150, 128)     147584    
_________________________________________________________________
block2_pool (MaxPooling2D)   (None, 75, 75, 128)      